### PR TITLE
fix timeout when stopping KVM machine with CRI-O container runtime

### PR DESCRIPTION
**Pull request by prezha**
we noticed in our integration tests that kvm machine with cri-o container runtime will sometime not gracefully shutdown waiting a given time of two minutes (eg: https://storage.googleapis.com/minikube-builds/logs/19736/36442/KVM_Linux_crio.html#fail_TestStartStop%2fgroup%2fembed-certs%2fserial%2fStop)

there is a known upstream [issue](https://github.com/cri-o/cri-o/issues/8609) that is tracking that

as we had similar issues in the past (and increased the timeout value several times), this pr aims to solve that problem by:
- waiting initial 90 seconds for the kvm machine to gracefully stop, and if that fails, instead of terminating with error:
- requesting a forceful shutdown (ie, "destroy" in libvirt) and waiting for another 30 seconds

in total, this should: self-resolve the issue in many cases, while reducing the total wait period

bonus:
- ensure any error from trying to close connection/domain is caught and logged but does not overwrite the underlying error
- more consistent error handling
- more consistent logging

### CODE REVIEW

**Comment by spowelljr** [comment]
Hi @prezha, following up on this according to https://github.com/cri-o/cri-o/issues/8609#issuecomment-2408033958, this was fixed in [crun 1.17](https://github.com/containers/crun/releases/tag/1.17). We've updated to crun 1.17 in https://github.com/kubernetes/minikube/pull/19640, is this PR still needed?

**Comment by prezha** [comment]
hey @spowelljr, i think we do as we've occasionally seen these timeouts even before (for reasons related to other/different issues), so it would also reduce the flakiness of our integration tests

File: pkg/drivers/kvm/kvm.go [hunk 1]
```diff
@@ -236,57 +247,79 @@ func (d *Driver) DriverName() string {
...
+	if err := d.ensureNetwork(); err != nil {
 		return errors.Wrap(err, "ensuring active networks")
 	}
 
-	log.Info("Getting domain xml...")
+	log.Info("getting domain XML...")
 	dom, conn, err := d.getDomain()
 	if err != nil {
-		return errors.Wrap(err, "getting connection")
+		return errors.Wrap(err, "getting domain")
```
Comment (spowelljr) [inline pkg/drivers/kvm/kvm.go:302]:
```suggestion
		return errors.Wrap(err, "getting domain XML")
```

File: pkg/drivers/kvm/kvm.go [hunk 2]
```diff
@@ -297,44 +330,48 @@ func (d *Driver) waitForStaticIP(conn *libvirt.Connect) error {
...
 		return errors.Wrap(err, "creating network")
 	}
+
 	if d.GPU {
-		log.Info("Creating devices...")
-		d.DevicesXML, err = getDevicesXML()
+		log.Info("creating devices...")
+		xml, err := getDevicesXML()
 		if err != nil {
 			return errors.Wrap(err, "creating devices")
```
Comment (spowelljr) [inline pkg/drivers/kvm/kvm.go:372]:
```suggestion
		log.Info("getting devices XML...")
		xml, err := getDevicesXML()
		if err != nil {
			return errors.Wrap(err, "getting devices XML")
```

File: pkg/drivers/kvm/kvm.go [hunk 3]
```diff
@@ -346,22 +383,22 @@ func (d *Driver) Create() (err error) {
...
+	if err := pkgdrivers.MakeDiskImage(d.BaseDriver, d.Boot2DockerURL, d.DiskSize); err != nil {
+		return errors.Wrap(err, "creating disk")
 	}
 
 	if d.ExtraDisks > 20 {
 		// Limiting the number of disks to 20 arbitrarily. If more disks are
 		// needed, the logical name generation has to changed to create them if
 		// the form hdaa, hdab, etc
-		return errors.Wrap(err, "cannot create more than 20 extra disks")
+		return errors.New("creating more than 20 extra disks")
```
Comment (spowelljr) [inline pkg/drivers/kvm/kvm.go:401]: The before message is a better user facing error, `cannot create` vs `creating`, although we should have validation for this way before

**Review by spowelljr — CHANGES_REQUESTED** [review]

File: pkg/drivers/kvm/kvm.go (hunk 3 above)
Comment (prezha) [reply pkg/drivers/kvm/kvm.go]: agree, reworded it

File: pkg/drivers/kvm/kvm.go [hunk 4]
```diff
@@ -427,100 +468,154 @@ func ensureDirPermissions(store string) error {
...
+		if s == state.Stopped {
+			return state.Stopped, nil
 		}
 
+		log.Infof("waiting for domain %s %d/%d", method, i, maxsec)
+		time.Sleep(1 * time.Second)
 	}
 
-	return fmt.Errorf("unable to stop vm, current state %q", s.String())
+	return s, fmt.Errorf("timed out waitig for domain %s, current state is %q", method, s)
```
Comment (spowelljr) [inline pkg/drivers/kvm/kvm.go:556]:
```suggestion
	return s, fmt.Errorf("timed out waiting for domain %s, current state is %q", method, s)
```

**Review by spowelljr — APPROVED** [review]
Found a typo, but the tests look good!

**Comment by medyagh** [comment]
thank you @prezha for improving the KVM, its nice to see less failed tests

**Pull request merged**